File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        TreeNode* cur=root;
        while(cur){
            if(!cur->left){
                k--;
                if(k==0)return cur->val;
                cur=cur->right;
            }else{
                TreeNode* pred=cur->left;
                while(pred->right && pred->right!=cur)pred=pred->right;

                if(!pred->right){
                    pred->right=cur;
                    cur=cur->left;
                }else {
                    pred->right=nullptr;
                    k--;
                    if(k==0)return cur->val;
                    cur=cur->right;
                }
            }
        }
        return -1;
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def dfs(self,root):
        if root is None:
            return 0
        left=self.dfs(root.left)
        right=self.dfs(root.right)
        self.maxi=max(self.maxi,left+right)

        return 1+max(left,right)

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.maxi=-1
        self.dfs(root)
        return self.maxi

        